src/Database/Type.php

```diff
@@ -106,21 +106,23 @@ public static function build($name)
             throw new InvalidArgumentException(sprintf('Unknown type "%s"', $name));
         }
         if (is_string(static::$_types[$name])) {
-            return static::$_builtTypes[$name] = new static::$_types[$name]($name);
+            static::$_types[$name] = new static::$_types[$name]($name);
```

Comment:
$_types[$name] just have  string values 
do you want to save object in here ?

Reply:
It always expects that it can contain objects. See map() with array as first arg.

Reply:
I see
Not expect
Arrays of string `string[]` not arrays of objects

Reply:
We can also deprecate the \Cake\Database\Type[] part of map() as array setter and force the _types to always be of type string[] in future versions.
I am fine either way. But as of currently, the above tweak seems to re-use an object if possible instead of re-new it. More like nano-opt probably.

Reply:
That use once per key
See line 102 an 114
I pretty sure that use just once and not use again
Except after clear()

Reply:
So type of string is true

Reply:
True, we could sure deprecate the object assignment for map($arrayOfTypes) then.

Reply:
1) I test your branch on local 
```php
vendor/bin/phpunit tests/TestCase/Database/Type
PHPUnit 5.7.21 by Sebastian Bergmann and contributors.

...............................................................  63 / 125 ( 50%)
.......................................................S......  125 / 125 (100%)

Time: 204 ms, Memory: 11.25MB

OK, but incomplete, skipped, or risky tests!
Tests: 125, Assertions: 279, Skipped: 1.
```
2) I not find any usage of object for type() in tests

Reply:
1) I test your branch on local
`vendor/bin/phpunit tests/TestCase/Database`
all test :+1: 
2) I not find any usage of object for type() in tests

Reply:
I didnt say there are tests for it, thats also why I added them.
I said that the code itself already expects them to be possibly of type object. So in light of that this only makes it official and obvious what the code always has already been checking for.

Reply:
@dereuromark  
:+1:  you are right
But I believe is irrational to accept object on `$_types`, because we have `$_builtTypes` for that

Reply:
That also concerned me, yeah. We should really deprecate that part and remove it in 4.0.